upload/library/Sidane/Threadmarks/DataWriter/Threadmark.php

```diff
@@ -0,0 +1,136 @@
+<?php
+
+class Sidane_Threadmarks_DataWriter_Threadmark extends XenForo_DataWriter
+{
+
+  protected function _getFields()
+  {
+    return array(
+      'threadmarks' => array(
+        'threadmark_id'          => array('type' => self::TYPE_UINT, 'autoIncrement' => true),
+        'user_id'                => array('type' => self::TYPE_UINT, 'required' => true),
+        'post_date'              => array('type' => self::TYPE_UINT, 'required' => true, 'default' => XenForo_Application::$time),
+        'thread_id'              => array('type' => self::TYPE_UINT, 'required' => true),
+        'post_id'                => array('type' => self::TYPE_UINT, 'required' => true),
+        'label'                  => array('type' => self::TYPE_STRING, 'required' => true, 'maxLength' => 75,
+             'requiredError' => 'please_enter_label_for_threadmark'
+        ),
+        'message_state'          => array('type' => self::TYPE_STRING, 'default' => 'visible',
+          'allowedValues' => array('visible', 'moderated', 'deleted')
+        ),
+        'last_edit_date'         => array('type' => self::TYPE_UINT, 'default' => 0),
+        'last_edit_user_id'      => array('type' => self::TYPE_UINT, 'default' => 0),
+        'edit_count'             => array('type' => self::TYPE_UINT_FORCED, 'default' => 0),
+        'position'               => array('type' => self::TYPE_UINT_FORCED),
```

Comment:
Position + message_state permit more efficient displaying of recent threadmarks, and much more efficient displaying of all threadmarks as all ordering information is now stored in the threadmark table.